--- sudo-1.9.17p2.orig/config.h.in
+++ sudo-1.9.17p2/config.h.in
@@ -1705,4 +1705,9 @@
 # define NO_LEAKS
 #endif
 
+/* QNX compat */
+#if defined(__QNX__) && !defined(SA_RESTART)
+# define SA_RESTART 0
+#endif
+
 #endif /* SUDO_CONFIG_H */
--- sudo-1.9.17p2.orig/lib/util/getentropy.c
+++ sudo-1.9.17p2/lib/util/getentropy.c
@@ -67,7 +67,11 @@
 # include <sys/auxv.h>
 #endif
 #ifdef HAVE_DL_ITERATE_PHDR
-# include <link.h>
+# ifdef __QNX__
+#  include <sys/link.h>
+# else
+#  include <link.h>
+# endif
 #endif
 #ifdef HAVE_OPENSSL
 # if defined(HAVE_WOLFSSL)
--- sudo-1.9.17p2.orig/lib/util/mkdir_parents.c
+++ sudo-1.9.17p2/lib/util/mkdir_parents.c
@@ -94,9 +94,9 @@
 	do {
 	    cp++;
 	} while (*cp == '/');
-	parentfd = open("/", O_RDONLY|O_NONBLOCK);
+	parentfd = open("/", O_RDONLY|O_NONBLOCK|O_DIRECTORY);
     } else {
-	parentfd = open(".", O_RDONLY|O_NONBLOCK);
+	parentfd = open(".", O_RDONLY|O_NONBLOCK|O_DIRECTORY);
     }
     if (parentfd == -1) {
 	if (!quiet)
@@ -123,7 +123,7 @@
 	memcpy(name, cp, len);
 	name[len] = '\0';
 reopen:
-	dfd = openat(parentfd, name, O_RDONLY|O_NONBLOCK, 0);
+	dfd = openat(parentfd, name, O_RDONLY|O_NONBLOCK|O_DIRECTORY, 0);
 	if (dfd == -1) {
 	    if (errno != ENOENT) {
 		if (!quiet) {
@@ -133,7 +133,7 @@
 		goto bad;
 	    }
 	    if (mkdirat(parentfd, name, mode) == 0) {
-		dfd = openat(parentfd, name, O_RDONLY|O_NONBLOCK|O_NOFOLLOW, 0);
+		dfd = openat(parentfd, name, O_RDONLY|O_NONBLOCK|O_NOFOLLOW|O_DIRECTORY, 0);
 		if (dfd == -1) {
 		    if (!quiet) {
 			sudo_warn(U_("unable to open %.*s"),
--- sudo-1.9.17p2.orig/lib/util/secure_path.c
+++ sudo-1.9.17p2/lib/util/secure_path.c
@@ -126,7 +126,7 @@
     if (sb == NULL)
 	sb = &stat_buf;
 
-    fd = open(path, O_RDONLY|O_NONBLOCK);
+    fd = open(path, O_RDONLY|O_NONBLOCK|O_DIRECTORY);
     if (fd == -1 || fstat(fd, sb) != 0) {
 	if (fd != -1)
 	    close(fd);
--- sudo-1.9.17p2.orig/pathnames.h.in
+++ sudo-1.9.17p2/pathnames.h.in
@@ -192,7 +192,7 @@
 #endif /* _PATH_ASAN_LIB */
 
 #ifndef _PATH_VI
-# undef _PATH_VI
+# define _PATH_VI	"/bin/vi"
 #endif /* _PATH_VI */
 
 #ifndef _PATH_MV
--- sudo-1.9.17p2.orig/plugins/sudoers/boottime.c
+++ sudo-1.9.17p2/plugins/sudoers/boottime.c
@@ -33,6 +33,9 @@
 #include <limits.h>
 #include <time.h>
 #ifndef __linux__
+# if defined(__QNX__)
+# include <utmp.h>
+# endif
 # if defined(HAVE_SYS_SYSCTL_H)
 #  include <sys/sysctl.h>
 # elif defined(HAVE_GETUTXID)
--- sudo-1.9.17p2.orig/plugins/sudoers/policy.c
+++ sudo-1.9.17p2/plugins/sudoers/policy.c
@@ -992,10 +992,12 @@
         if ((command_info[info_len++] = sudo_new_key_val("rlimit_data", def_rlimit_data)) == NULL)
             goto oom;
     }
+#ifndef __QNX__
     if (def_rlimit_fsize != NULL) {
         if ((command_info[info_len++] = sudo_new_key_val("rlimit_fsize", def_rlimit_fsize)) == NULL)
             goto oom;
     }
+#endif
     if (def_rlimit_locks != NULL) {
         if ((command_info[info_len++] = sudo_new_key_val("rlimit_locks", def_rlimit_locks)) == NULL)
             goto oom;
--- sudo-1.9.17p2.orig/plugins/sudoers/timestamp.c
+++ sudo-1.9.17p2/plugins/sudoers/timestamp.c
@@ -217,10 +217,10 @@
 	    if (!quiet)
 		sudo_warn(U_("unable to mkdir %s"), path);
 	} else {
-	    fd = openat(parentfd, base, O_RDONLY|O_NONBLOCK, 0);
+	    fd = openat(parentfd, base, O_RDONLY|O_NONBLOCK|O_DIRECTORY, 0);
 	    if (fd == -1) {
 		sudo_debug_printf(SUDO_DEBUG_ERROR|SUDO_DEBUG_ERRNO,
-		    "%s: unable to open %s", __func__, path);
+		    "%s: bruh unable to open %s", __func__, path);
 	    } else if (fchown(fd, owner, group) != 0) {
 		sudo_debug_printf(SUDO_DEBUG_ERROR|SUDO_DEBUG_ERRNO,
 		    "%s: unable to chown %d:%d %s", __func__,
@@ -314,7 +314,7 @@
 
     if (timestamp_uid != 0)
 	uid_changed = set_perms(NULL, PERM_TIMESTAMP);
-    fd = openat(dfd, path, flags, S_IRUSR|S_IWUSR);
+    fd = openat(dfd, path, flags, S_IRUSR|S_IWUSR|O_DIRECTORY);
     if (uid_changed && !restore_perms()) {
 	/* Unable to restore permissions, should not happen. */
 	if (fd != -1) {
@@ -1033,7 +1033,7 @@
     }
 #endif
 
-    dfd = open(def_timestampdir, O_RDONLY|O_NONBLOCK);
+    dfd = open(def_timestampdir, O_RDONLY|O_NONBLOCK|O_DIRECTORY);
     if (dfd == -1) {
 	if (errno != ENOENT)
 	    ret = -1;
--- sudo-1.9.17p2.orig/src/get_pty.c
+++ sudo-1.9.17p2/src/get_pty.c
@@ -43,6 +43,8 @@
 #  include <util.h>		/* macOS */
 # elif defined(HAVE_PTY_H)
 #  include <pty.h>		/* Linux */
+# elif defined(__QNX__)
+#  include <unix.h>		/* QNX */
 # else
 #  include <termios.h>		/* Solaris */
 # endif
--- sudo-1.9.17p2.orig/src/limits.c
+++ sudo-1.9.17p2/src/limits.c
@@ -129,7 +129,7 @@
     {
 	"rlimit_fsize",
 	RLIMIT_FSIZE,
-	true,					/* override */
+	false,					/* override */
 	false,					/* saved */
 	false,					/* policy */
 	false,					/* preserve */
--- sudo-1.9.17p2.orig/src/utmp.c
+++ sudo-1.9.17p2/src/utmp.c
@@ -55,7 +55,16 @@
 /*
  * Simplify handling of different utmp types.
  */
-#if defined(HAVE_GETUTSID)
+#if defined(__QNX__)
+# define sudo_getutline(u)	getutline(u)
+struct utmp *sudo_pututline(struct utmp *utmp)
+{
+       pututline(utmp);
+       return utmp;
+}
+# define sudo_setutent()	setutent()
+# define sudo_endutent()	endutent()
+#elif defined(HAVE_GETUTSID)
 # define sudo_getutline(u)	GETUTSLINE(u)
 # define sudo_pututline(u)	PUTUTSLINE(u)
 # define sudo_setutent()	SETUTSENT()
